### PR TITLE
Support yield keyword

### LINKED ISSUE #4309: Don't recognize generators
```
#![feature(generators)]

fn main() {
    let x = || { yield 3 };
}
```

When using VSCode with this code I notice that:
1. It seems a ```yield``` with argument is reported by analyzer as error.
2. Also helping hint about ```x``` type displays that ```x``` have closure type ```|| -> i32``` instead of generator trait type.

### PR DESCRIPTION
**Pull request by sasurau4**
Part of #4309 
The inference of yield will be implemented at another PR.

### CODE REVIEW

**Comment by Veykril** [comment]
Those files are auto-generated by xtask. To support yield we have to add a `YieldExpr` rule to the [ungrammar](https://github.com/rust-analyzer/ungrammar/blob/master/rust.ungram) file and the [Expr](https://github.com/rust-analyzer/ungrammar/blob/master/rust.ungram#L323-L353) rule. Then regenerate these files here and adjust the parser.

**Comment by lnicola** [comment]
`ungrammar 1.7` is out.

**Comment by Veykril** [comment]
As a heads up since you haven't touched that part here before either, to make the parser support the yield syntax you will have to add the parsing for it at the correct places in the https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/parser/src/grammar/expressions/atom.rs module.

File: xtask/src/ast_src.rs [hunk 1]
```diff
@@ -67,8 +67,8 @@ pub(crate) const KINDS_SRC: KindsSrc = KindsSrc {
     keywords: &[
         "as", "async", "await", "box", "break", "const", "continue", "crate", "dyn", "else",
         "enum", "extern", "false", "fn", "for", "if", "impl", "in", "let", "loop", "macro",
-        "match", "mod", "move", "mut", "pub", "ref", "return", "self", "static", "struct", "super",
-        "trait", "true", "try", "type", "unsafe", "use", "where", "while",
+        "match", "mod", "move", "mut", "pub", "ref", "return", "yield", "self", "static", "struct",
```
Comment (lnicola) [inline xtask/src/ast_src.rs:70]: I think these were sorted.

**Comment by sasurau4** [comment]
@Veykril  Thanks for guiding!

File: crates/hir_ty/src/infer/expr.rs [hunk 2]
```diff
@@ -358,7 +358,7 @@ impl<'a> InferenceContext<'a> {
 
                 Ty::simple(TypeCtor::Never)
             }
-            Expr::Return { expr } => {
+            Expr::Return { expr } | Expr::Yield { expr } => {
                 if let Some(expr) = expr {
                     self.infer_expr_coerce(*expr, &Expectation::has_type(self.return_ty.clone()));
```
Comment (Veykril) [inline crates/hir_ty/src/infer/expr.rs:363]: This doesn't apply to yield I think. A generator may have a return type and a yield type that aren't the same. To properly implement inference for `yield` we need to track the yield type in the `InferenceContext` I believe but that can be done in a separate PR I'd say.
So for now we probably want something like this:
```rust
Expr::Yield { expr } => {
	  // FIXME: track yield type for coercion
	  if let Some(expr) = expr {
	      self.infer_expr(expr, &Expectation::none());
	  }
	  Ty::simple(TypeCtor::Never)
}
```

**Comment by Veykril** [comment]
lgtm 👍 
bors r+

**Pull request merged**